### ISSUE TITLE
Jellyfin on Truenas - ffmpeg problem

### ISSUE THREAD

**Issue by brsheron** [Thefrank/jellyfin-server-freebsd#57]
Hi.
I'm not sure if this is the right place to report this, so if it isn't, please let me know.
I recently tried to update my jellyfin instance on TrueNAS via the GUI. It didn't go well so I deleted it and reinstalled it.
Previously I could play my media and I could see ffmpeg running doing some sort of re-encoding?
Anyway, after the reinstall it tells me that the media is in the wrong format.

This is the error message in the GUI:

_"Playback Error
This client isn't compatible with the media and the server isn't sending a compatible media format."_

In the logfile it looks like this:
[2023-08-12 22:48:58.592 +02:00] [ERR] [46] Jellyfin.Server.Middleware.ExceptionMiddleware: Error processing request. URL "GET" "/videos/<redacted>/live.m3u8".
System.ArgumentException: FFmpeg path not set.
   at Jellyfin.Api.Helpers.TranscodingJobHelper.StartFfMpeg(StreamState state, String outputPath, String commandLineArguments, HttpRequest request, TranscodingJobType transcodingJobType, CancellationTokenSource cancellationTokenSource, String workingDirectory)
   at Jellyfin.Api.Controllers.DynamicHlsController.GetLiveHlsStream(Guid itemId, String container, Nullable`1 static, String params, String tag, String deviceProfileId, String playSessionId, String segmentContainer, Nullable`1 segmentLength, Nullable`1 minSegments, String mediaSourceId, String deviceId, String audioCodec, Nullable`1 enableAutoStreamCopy, Nullable`1 allowVideoStreamCopy, Nullable`1 allowAudioStreamCopy, Nullable`1 breakOnNonKeyFrames, Nullable`1 audioSampleRate, Nullable`1 maxAudioBitDepth, Nullable`1 audioBitRate, Nullable`1 audioChannels, Nullable`1 maxAudioChannels, String profile, String level, Nullable`1 framerate, Nullable`1 maxFramerate, Nullable`1 copyTimestamps, Nullable`1 startTimeTicks, Nullable`1 width, Nullable`1 height, Nullable`1 videoBitRate, Nullable`1 subtitleStreamIndex, Nullable`1 subtitleMethod, Nullable`1 maxRefFrames, Nullable`1 maxVideoBitDepth, Nullable`1 requireAvc, Nullable`1 deInterlace, Nullable`1 requireNonAnamorphic, Nullable`1 transcodingMaxAudioChannels, Nullable`1 cpuCoreLimit, String liveStreamId, Nullable`1 enableMpegtsM2TsMode, String videoCodec, String subtitleCodec, String transcodeReasons, Nullable`1 audioStreamIndex, Nullable`1 videoStreamIndex, Nullable`1 context, Dictionary`2 streamOptions, Nullable`1 maxWidth, Nullable`1 maxHeight, Nullable`1 enableSubtitlesInManifest)
   at lambda_method1568(Closure , Object )
   at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.TaskOfActionResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Awaited|12_0(ControllerActionInvoker invoker, ValueTask`1 actionResultValueTask)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
--- End of stack trace from previous location ---
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()

Looks like there's a problem with the ffmpeg path?

However if I try to set the ffmpeg path in the GUI I get this error message. I can see the ffmpeg binary in the directory I'm selecting. Doesn't work with pointing to the directory, or the binary itself.
_"We're unable to find FFmpeg using the path you've entered. FFprobe is also required and must exist in the same folder. These components are normally bundled together in the same download. Please check the path and try again."_

However if I then go into the jail and try to just check the ffmpeg binary it looks like this:
./ffmpeg
ld-elf.so.1: /usr/local/bin/libtasn1.so.6: Undefined symbol "strverscmp@FBSD_1.7

Is something broken? Can I fix it somehow? Any help appreciated.

**Comment by Thefrank** [comment]
In the jail, can you:
 - `freebsd-version -ru; uname -a`
 - `pkg -vv | grep ^OSVERSION`

TrueNAS console:
 - `freebsd-version -kru -j (JAIL_NUMBER)`
 JAIL_NUMBER should be the number associated with the jail that runs jellyfin

The symbol version `FBSD_1.7` is listed for 14.0-CURRENT so it sounds like a mismatch between the version of TrueNAS, the jail, and the packages for the jail but we will find out more from the commands above.

**Comment by brsheron** [comment]
freebsd-version -kru -j reports this:
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/92d9400b-347b-4249-b0ae-58f91cb3be93)
13.1-RELEASE-p7
13.1-RELEASE-p7
13.1-RELEASE-p7
13.1-RELEASE-p9

In the jail:
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/d47f0134-eb49-4cda-9a3a-645132d738a9)
13.1-RELEASE-p7
13.1-RELEASE-p9
The uname is too long for me to type, but the screenshot should show it.

**Comment by Thefrank** [comment]
So there does appear to be a mismatch between the Jail and the Packages used by it. Try and update the jail and see if that resolves it

**Comment by brsheron** [comment]
I tried updating the jail both through the web-gui in TrueNAS and the manual method, but neither seems to update the jail. 
This is the output from the manual upgrade.
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/c68a44e0-1123-4ee8-b711-ebaf415cbd55)


The only "error" I can see is perhaps this one? And I'm not even sure it's an error.
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/89a37007-40b8-49a2-aaef-60f0def4c0d6)

I'm still very much in the dark here and any help appreciated.

**Comment by Thefrank** [comment]
Do the two lines returned from `freebsd-version -ru` match? if yes, then hopefully everything is fixed. If the DO match but things are still not working try
```
pkg bootstrap -f
pkg upgrade
```
in the jail

**Comment by brsheron** [comment]
In the jail there's a slight difference between the versions.
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/a30a1cb2-48ac-40b4-ad90-579cca853ada)

However outside, there doesn't seem to be.
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/3d7516b0-8718-4c78-90a2-81ecda1dce31)

Again, in the jail, doing the pkg bootstrap thing...
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/ba96d0f6-7006-41e2-a13b-c0ae3ea9d21b)

Doesn't seem to change anything, right?
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/703fd956-6126-481c-a76e-446908ef8fc7)

I've run the pkg bootstrap and upgrade several times, and I get this message every time. 
![image](https://github.com/Thefrank/jellyfin-server-freebsd/assets/120510077/b326e213-6a4f-4bbd-842c-5581eed75003)

So I don't think the bootstrap thing does anything. ffmpeg still won't work.
Anything else I can try? I can reinstall if that helps.

**Comment by Thefrank** [comment]
See that big warning about "Newer FreeBSD version for package pkg"? That is an indication that something is wrong between the package version and the kernel version. `pkg` and everything that it will pull down is built for a different kernel version than you are running.

Generally, jails should not have a version higher than the base system and most jail managers either outright prevent that or warn against it even during update/upgrade. Under normal FreeBSD this is somewhat easy, but slow, to fix; easier and faster with snapshots. TrueNAS... no idea.

You can see if there is an older cached version still on your system and install that
```
find /var/cache/pkg/ -name "ffmpeg*"
```
then
`pkg install /var/cache/pkg/OLDVERSION~HASH.pkg`
you might need to `-f`

Alternatively, I have a statically built version of FFmpeg here: https://github.com/Thefrank/ffmpeg-static-freebsd

Random concern: why is something from the FreeBSD:13 ports branch including a FreeBSD 14+ symbol (`FBSD_1.7`)?

**Comment by brsheron** [comment]
I think I get what you're saying and I'll probably end up using the static version. Thanks for that.
I'm somewhat confused still about where the problem actually is.
Is it the creator of the jail that's asking for the wrong version? The TrueNAS guys who included (or pointed) to a wrong version of ffmpeg? Someone else?
I'd like to report this to someone and perhaps have it fixed, but I don't know where to turn.

**Comment by Thefrank** [comment]
Using both FreeBSD 12.4 and 13.2 on `pkg`'s "quarterly" branch I am not seeing a broken `ffmpeg`. 
I was only able to reproduce this in TrueNAS because it is still on 13.1:
-> old 13.1-p1 on "latest" pkg branch
-> update with iocage to 13.1-p9
-> remove all packages, delete cache
**-> `pkg upgrade` and ignore big warning**
-> reinstall ffmpeg
-> `ld-elf.so.1: /usr/local/lib/libtasn1.so.6: Undefined symbol "strverscmp@FBSD_1.7"`

Odd...Time to dig!

13.1 is EOL. 
13.2 added  `strcmp`  https://cgit.freebsd.org/src/commit/?h=releng/13.2&id=27e60668bf29dc4da3409168967974e7ee3ffddc.  The version warning from `pkg` was important here
`jellyfin` and my static version of `ffmpeg` should both work fine on TrueNAS as they were built against older versions of FreeBSD (13.1)

**Comment by brsheron** [comment]
I think I might have solved it by not using the "plugin" provided by TrueNAS but instead create a jail manually and install jellyfin from there. Now everything seems to work as it should. 

Probably what I should have done from the start instead of using the plugin system, which seems to be broken.

Thanks for all your help guys. You can close this as solved.

**Comment by TNAS123** [comment]
Hi @brsheron , could I ask you for the steps you took to resolve this issue? As you can see I've raised an issue for the exact same. Currently this is breaking my whole Jellyfin media server so any step by step help how you resolved it would be appreciated. Thanks

**Comment by brsheron** [comment]
Hi.
I'm afraid I didn't solve it. I had a gordian knot moment.

What I did was that I created a new jail (not through the plugin system) and chose the latest available version of the os (13.2?), installed jellyfin and moved my media files over to the new jail and deleted the plugin-jail.

Not elegant, but it worked.

**Comment by spz2k9** [comment]
The manual installation will also create a 13.2 jail.
